posthog/temporal/weekly_digest/activities.py

```diff
@@ -0,0 +1,392 @@
+        bind_contextvars(digest_key=input.digest_key, batch_start=input.batch[0], batch_end=input.batch[1])
+        logger = LOGGER.bind()
+        logger.info("Generating organization-level digest batch")
+
+        organization_count = 0
+        team_count = 0
+
+        r = await redis.from_url(f"redis://{input.redis_host}:{input.redis_port}?decode_responses=true")
+
+        batch_start, batch_end = input.batch
+        async for organization in query_orgs_for_digest()[batch_start:batch_end]:
+            organization_count += 1
+
+            team_digests: list[TeamDigest] = []
+
+            async for team in query_org_teams(organization):
+                team_count += 1
+
+                digest_data: list[str] = await asyncio.gather(
+                    *[
+                        r.get(f"{input.digest_key}-dashboards-{team.id}"),
+                        r.get(f"{input.digest_key}-event-definitions-{team.id}"),
+                        r.get(f"{input.digest_key}-experiments-launched-{team.id}"),
+                        r.get(f"{input.digest_key}-experiments-completed-{team.id}"),
+                        r.get(f"{input.digest_key}-external-data-sources-{team.id}"),
+                        r.get(f"{input.digest_key}-surveys-launched-{team.id}"),
+                        r.get(f"{input.digest_key}-feature-flags-{team.id}"),
+                    ]
+                )
+
+                team_digests.append(
+                    TeamDigest(
+                        id=team.id,
+                        name=team.name,
+                        dashboards=DashboardList.model_validate_json(digest_data[0]),
+                        event_definitions=EventDefinitionList.model_validate_json(digest_data[1]),
+                        experiments_launched=ExperimentList.model_validate_json(digest_data[2]),
+                        experiments_completed=ExperimentList.model_validate_json(digest_data[3]),
+                        external_data_sources=ExternalDataSourceList.model_validate_json(digest_data[4]),
+                        surveys_launched=SurveyList.model_validate_json(digest_data[5]),
+                        feature_flags=FeatureFlagList.model_validate_json(digest_data[6]),
+                    )
+                )
+
+            org_digest = OrganizationDigest(
+                id=organization.id,
+                name=organization.name,
+                created_at=organization.created_at,
+                team_digests=team_digests,
+            )
+
+            key: str = f"{input.digest_key}-{organization.id}"
+            await r.set(key, org_digest.model_dump_json(), ex=input.redis_ttl)
+
+        await r.aclose()
+
+        logger.info(
+            "Finished generating organization-level digests",
+            organization_count=organization_count,
+            team_count=team_count,
+        )
+
+
+@activity.defn(name="send-weekly-digest-batch")
+async def send_weekly_digest_batch(input: SendWeeklyDigestBatchInput) -> None:
```

Comment:
nit pick that these are all pretty similar
would be easier to see the differences if they were using one or more extracted methods
but 🤷